Add bytecode profiler

jmerle opened the pull request:
This PR adds a bytecode profiler to the engine and a visualizer of the recorded data to the client. The visualizer uses [speedscope](https://github.com/jlfwong/speedscope) to display the results. Here's a little demo: [https://www.youtube.com/watch?v=j5noBI3exaQ](https://www.youtube.com/watch?v=j5noBI3exaQ).

This PR is accompanied by https://github.com/battlecode/battlecode20-scaffold/pull/9 because the profiler is toggled by a Gradle property. This is also written out on the Help tab of the sidebar.

The bytecode profiling _does not_ affect actual bytecode usage, the entire profiling part is wrapped around the bytecode counting so the profiling isn't seen as player code.

Recorded data is stored in the bc20 file, for which I extended the MatchFooter table in the FlatBuffers schema. bc20 files with profiling data can grow somewhat big (I've seen 10+ mb playing my sprint bot against itself), but since it's not supposed to be ran in scrimmages/tournaments that shouldn't be a problem. What also helps is that the profiler only profiles player methods, so if RobotController.run() calls a method on (for example) an ArrayList, the call to ArrayList is not monitored and the bytecode used is simply added to the record of the RobotController.run method.

arvid220u commented on client/visualizer/electron-main.js:
```diff
@@ -52,6 +52,11 @@ function createWindow () {
+// This is needed so that the speedscope iframe on the Profiler tab does not throw cross-origin errors
+// The speedscope iframe loads speedscope using the file:// protocol which considers all URLs to have separate origins
+// The only way to get around this is to tell Chrome to allow it anyways
+electron.app.commandLine.appendSwitch('disable-site-isolation-trials');
```
this seems less than ideal but maybe it is inevitable

(just commenting so i remember to go back to this)

jmerle replied:
When this line is removed, the Profiler tab works when the application is first started but breaks when the user hits Ctrl+R. There may be other solutions to this, but it already took me 30 minutes to fix it using this line and I also had a sprint bot to work on :)

arvid220u replied:
oh i see, let's just leave this here then

arvid220u commented on client/visualizer/README.md:
```diff
@@ -54,7 +54,7 @@ $ npm run prod
 
 - Mount the assets in `bc20` at `/bc20/` on your webserver.
-- `<script src="/bc20/bundle.js"></script>`
+- `<script src="/bc20/app.js"></script>`
```
check this

arvid220u commented on client/visualizer/index.html:
```diff
@@ -6,7 +6,7 @@
     <body>
         <div id='client'/>
-        <script type="text/javascript" src="bc20/bundle.js" charset="utf-8"></script>
+        <script type="text/javascript" src="bc20/app.js" charset="utf-8"></script>
```
bundle.js app.js change

arvid220u commented on client/visualizer/package.json:
```diff
@@ -57,7 +58,7 @@
     "compression": "maximum",
     "files": [
-      "bc20/*",
+      "bc20/**/*",
```
hmm

jmerle replied:
The CopyWebpackPlugin now copies the speedscope code to "bc20/speedscope". "bc20/*" non-recursively takes all contents inside the "bc20/" directory and skips the files inside "bc20/speedscope/" because that's a different directory. I might be wrong here though.

arvid220u commented on client/visualizer/src/game/gamearea/gamearea.ts:
```diff
@@ -125,17 +127,22 @@ export default class GameArea {
     if(shouldListen) {
       window.onresize = function() {
-        var wrapper = <HTMLDivElement> document.getElementById("canvas-wrapper")
+        var wrapper = <HTMLDivElement> document.getElementById("canvas-wrapper");
```
:0

jmerle replied:
I use IntelliJ and it highlighted all the missing semicolons, so I added some as I read over them and eventually gave up because they are used wayyyy too inconsistent :P

arvid220u commented on client/visualizer/src/game/sidebar/matchrunner.ts:
```diff
@@ -38,10 +38,11 @@ export default class MatchRunner {
   private deselectAllMaps: HTMLButtonElement;
   private runMatch: HTMLButtonElement;
-  private refreshButton: HTMLButtonElement
+  private refreshButton: HTMLButtonElement;
```
:0

arvid220u commented on client/visualizer/src/main/sidebar.ts:
```diff
@@ -125,8 +129,13 @@ export default class Sidebar {
     <b class="blue">Keyboard Shortcuts</b><br>
     LEFT - Step Back One Turn<br>
     RIGHT - Step Forward One Turn<br>
+<<<<<<< HEAD
```
let's change this

jmerle replied:
Whoops, forgot that merge conflict, can fix in ~2 hours

arvid220u commented on client/visualizer/src/main/sidebar.ts:
```diff
@@ -301,8 +319,13 @@ export default class Sidebar {
     }
 
-    // this.cb();
+    if (this.cb !== undefined) {
```
check this

jmerle replied:
this.cb() is set to update the gamearea content based on the currently open tab. I guess this was commented out after removing the map editor (currently the replay viewer is always shown) and I added it back in in order to show the speedscope iframe when the Profiler tab is active. Not sure about the `!== undefined` check, but I'm guessing it caused an error without it.

arvid220u replied:
Makes sense, thanks

arvid220u commented on client/visualizer/src/static/css/style.css:
```diff
@@ -86,11 +101,17 @@ form {
 #gamearea canvas {
     max-height: calc(100vh - 75px);
-    max-width: calc(100vw - 320px);
+    max-width: calc(100vw - 345px);
```
.

jmerle replied:
Sidebar has been made bigger to make space for the Profiler tab. As a consequence, the gamearea canvas has been made smaller so it doesn't flow underneath the sidebar.

arvid220u commented on client/visualizer/webpack.config.js:
```diff
@@ -10,11 +10,12 @@ var conf = {
     path: path.resolve(__dirname, 'bc20'),
     publicPath: '/bc20/',
-    filename: 'bundle.js'
+    filename: '[name].js'
```
ohhhh i see

jmerle replied:
Sidebar has been made bigger to make space for the Profiler tab. As a consequence the gamearea canvas has been made smaller.

arvid220u commented on client/visualizer/webpack.config.js:
```diff
@@ -27,6 +28,26 @@ var conf = {
+  devServer: {
+    // Required to ensure the files copied by the CopyWebpackPlugin are copied when running the dev server
+    writeToDisk: filePath => filePath.includes('speedscope/')
+  },
```
hmm why is this necessary

jmerle replied:
When using webpack-dev-server (with `npm run watch` in the client directory) the CopyWebpackPlugin doesn't actually copy the speedscope files without this line because webpack-dev-server does everything in-memory. This causes problems because those speedscope files are needed by the iframe visible when the Profiler tab is active. By using `writeToDisk` we can tell webpack-dev-server to write certain files to disk even when running primarily in-memory (in this case all the speedscope files).

arvid220u replied:
i see i see

arvid220u commented on engine/src/main/battlecode/instrumenter/bytecode/InstrumentingMethodVisitor.java:
```diff
@@ -153,6 +152,9 @@ public void visitMaxs(int maxStack, int maxLocals) {
         boolean anyTryCatch = tryCatchBlocks.size() > 0;
 
+        // must be called before addDebugHandler() so debug methods are profiled properly
+        addEnterMethodHandler();
```
so i guess even if instrumenting is disabled we inject the instrumenter code?

(again this is only my commments so i can figure out what happens here)

jmerle replied:
Not anymore

arvid220u commented on engine/src/main/battlecode/instrumenter/bytecode/InstrumentingMethodVisitor.java:
```diff
@@ -185,6 +187,36 @@ private static AbstractInsnNode nextInstruction(AbstractInsnNode n) {
         return n;
     }
 
+    private void addEnterMethodHandler() {
```
for later

arvid220u commented on engine/src/main/battlecode/instrumenter/inject/RobotMonitor.java:
```diff
@@ -129,11 +138,16 @@ public static void incrementBytecodes(int numBytecodes) {
 
+            // the profiler isn't enabled in ranked matches, so the integer overflow isn't really a problem
+            if (profiler != null) {
+                profiler.incrementBytecodes(numBytecodes + bytecodesToRemove);
```
we probably still want to avoid the overflow

arvid220u commented on engine/src/main/battlecode/instrumenter/inject/RobotMonitor.java:
```diff
@@ -235,6 +249,32 @@ public static long getRandomSeed() {
+     *
+     * @param methodName the name of the method that is being entered
+     */
+    public static void enterMethod(String methodName) {
```
oh ok this seems fine

arvid220u commented:
sorry for the many comments, just want to read through everything. i really like this change though and we'll probably merge it tomorrow!

jmerle commented:
That's what comments are for :)

I've replied to some of them to make it more clear and will fix the merge conflicts in about ~2 hours.

jmerle commented:
I've fixed the merge conflicts, fixed the potential integer overflow and made it so that it doesn't instrument the player code with profiling related code when profiling is disabled.

arvid220u commented:
@jmerle Have you also been getting `Exception in thread "main" java.lang.OutOfMemoryError: Java heap space` when running games with the profiler? I'm running a game with a lot of robots though so maybe that is why

arvid220u commented:
Amazing work! Thank you so much for doing this. I've merged this and will push it out tonight.
